scripts/monitor/visualization/resource_monitoring.py

```diff
@@ -0,0 +1,596 @@
+        :return: plot
+        """
+        x = self.mem_requested_for_task(task).loc[:, "shard"]
+        y1 = self.mem_requested_for_task(task).loc[:, "runtime_mem_total_gb"].astype("float64", copy=True)
+        y2 = self.mem_requested_for_task(task).loc[:, "meta_mem_total_gb"].astype("float64", copy=True)
+        plt.plot(x, y1)
+        plt.plot(x, y2)
+        plt.legend(["Runtime Memory Usage(gb)", "Meta Memory Usage(gb)"])
+        plt.show()
+        # plt.savefig(f"{__name__}.pdf")
+        return plt.figure()
+
+    def task_input_size(self, task_name_or_index, shard_index=None) -> float:
+        """
+        Summation of a task's input file sizes
+        Use:
+        task_input_size(row index of task/shard), or task_input_size(task name, shard index)
+        :param task_name_or_index: task name or index (details above)
+        :param shard_index: shard index if task name is given
+        :return:
+        """
+        pos = 0
+        if shard_index is not None:
+            tasks = self.meta_runtime.loc[:, "runtime_task_call_name"]  # find row index
+            shards = self.meta_runtime.loc[:, "runtime_shard"]
+            for i in range(0, tasks.size):
+                if tasks[i] == task_name_or_index and shards[i] == shard_index:
+                    pos = i
+        else:
+            if isinstance(task_name_or_index, int):
+                pos = task_name_or_index  # if only task_name is supplied it's treated as the row index
+            else:
+                raise Exception(f"Need an integer row index but received {task_name_or_index}")
+        input_ds = pd.DataFrame(data=self.meta_runtime.loc[pos, "meta_inputs"])
+        return input_ds[input_ds['type'].astype(str) == "file"].loc[:, "value"].astype(float).sum()
+
+    # ''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''
+
+    def cpu_use_violin(self, instance_id: int) -> plt.Figure:
+        """
+        Violin plot for CPU use
+        :param instance_id: id of vm a shard is running on
+        :return: plot
+        """
+        by_id = self.metrics[self.metrics["metrics_instance_id"] == instance_id]\
+            .sort_values("metrics_timestamp").loc[:, "metrics_cpu_used_percent"]
+        data = list()
+        for i in range(len(by_id.iloc[0])):
+            data.append(pd.Series(list(e[i] for e in by_id)))
+        fig = plt.figure()
+        axis = fig.add_axes([0, 0, 1, 1])
+        _ = axis.violinplot(data, showmeans=True, showextrema=True, showmedians=True)
+        name = self.get_shard_from_instance_id(instance_id)
+        axis.set_xlabel("CPU core #")
+        axis.set_ylabel("CPU % used")
+        axis.set_title(f"CPU Usage for {name[0]}, shard {name[1]}")
+        plt.show()
+        # plt.savefig(f"{self.cpu_use_violin.__name__}.pdf", bbox_inches='tight')
+        return fig
+
+    def get_instance_id(self, task: str, shard_idx: int) -> numpy.int64:
+        """
```

Comment:
How about expanding this a bit, and make it so it also supports unsharded tasks?

Reply:
it actually does, but you just have to use -1 as the index. I'll mention that in the description.